### PR TITLE
Add in dsh project hook ability

### PR DESCRIPTION
**Pull request by singularo**
* Re-order code in dsh_start a little to allow vars to be overridden
* Add hooks got dsh_start, dsh_shell, dsh_stop, dsh_purge
* Fix indenting in dsh_shell
* Provide example dsh.project which re-implements lost linux specific tweaks.

### CODE REVIEW

**Comment by acbramley** [comment]
The idea of the simplified docker compose set up was to avoid these kinds of things (nginx network etc). I'm also still not sure why you need the SSH socket stuff when we're just mounting in our ssh key?

**Comment by singularo** [comment]
The default setup now in wcms-foundation doesn't actually appear to work for linux:

```🐳` root@x1:/code# ssh-add -l
Could not open a connection to your authentication agent.
🐳 root@x1:/code# exit
simon@x1:~/uni/ua-wcms-foundation$ git status
On branch develop
Your branch is up-to-date with 'origin/develop'.
nothing to commit, working tree clean
simon@x1:~/uni/ua-wcms-foundation$ 
```

But as @pingers and I mostly commit from the outside machine, he probably hasn't noticed, or maybe I'm doing something weird, but that was a fresh `git pull && composer install` and it doesn't work.

Nginx  was actually the bit I started it for, then had permission issues when I accidentally did composer update in the container and decided to pull the user stuff in as well.

**Comment by acbramley** [comment]
@singularo interesting, I don't use `ssh-add` at all in the container, and as you say I usually do git/composer operations outside of the container. However, running `robo theme:update` inside the container does work for me (on OSX but has a very similar setup).

**Comment by acbramley** [comment]
Ahhh there's your problem, you're using a passphrase on your key ;)

File: dsh.project.example [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+# Overridden dsh_shell function
+start_shell() {
+  if ! docker ps | grep "\s${PROJECT}_web_1" > /dev/null; then
+    notice "Project not running, starting."
+    dsh_start
+  fi
+
+  # Setup docker to run the shell as the current user, thus removing any
```
Comment (pingers) [inline dsh.project.example]: I definitely prefer running as my user on linux. Maybe this is just a linux thing though?

Comment (singularo) [reply dsh.project.example]: Quite possibly, the `dsh` changes are minimal, and enable you to extend it to do that. And yeah, things like the paths updated with `robo theme:update` then should still be owned by the user, not root (as it did when I tested it before). I suspect that nfs is hiding ownership issues from Mac users.

**Comment by singularo** [comment]
Yeah, just had the permissions problem on a project using the updates without this MR and its definitely a pain not having your key in there already unlocked.
And we don't want to have to use a key without a password imho.

**Comment by singularo** [comment]
Just encountered the permission issue with updating researcher profiles to wcms-foundation as well, would like to get this in.

**Comment by singularo** [comment]
Lets get er in.

**Comment by pingers** [comment]
Let's do the last bit of clean up:
- remove the host only networking mode from the linux docker-compose file

**Comment by singularo** [comment]
That means adding back the start nginx-proxy stuff as well, really.

File: dsh [hunk 2]
```diff
@@ -12,13 +12,18 @@ if [ $(version ${DOCKER_COMPOSE_VERSION}) -lt $(version "1.21.0") ]; then
...
 
-# Determine the OS type of the host machine.
-if [ "$(uname)" == "Darwin" ]; then
-  HOST_TYPE='mac'
-  export DOCKER_COMPOSE_FILE='docker-compose.osx.yml'
-else
+if [ ${GITLAB_CI:-false} == "true" ]; then
   HOST_TYPE='linux'
-  export DOCKER_COMPOSE_FILE='docker-compose.linux.yml'
+  export DOCKER_COMPOSE_FILE='docker-compose.ci.yml'
```
Comment (pingers) [inline dsh]: Where is this file?

**Comment by singularo** [comment]
I'm inclined to abandon this, and just keep the adjustments to the docker-compose.linux.yml.

**Comment by singularo** [comment]
Cancelling this

**Pull request closed without merging**